extensions/avro/src/main/java/com/hazelcast/jet/avro/AvroSerializerHooks.java

```diff
@@ -19,42 +19,157 @@
+import org.apache.avro.io.EncoderFactory;
 import org.apache.avro.util.Utf8;
 
+import javax.annotation.Nonnull;
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+
+import static com.hazelcast.jet.impl.util.ExceptionUtil.rethrow;
 
 /**
- * Hazelcast serializer hooks for the classes in the {@code
- * com.hazelcast.jet.avrp} package. This is not a public-facing API.
+ * Hazelcast serializer hooks for the classes in the {@code com.hazelcast.jet.avro} package.
  */
-class AvroSerializerHooks {
+public final class AvroSerializerHooks {
+    private AvroSerializerHooks() { }
+
+    public static <T> byte[] serialize(DatumWriter<T> datumWriter, T data) {
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+            BinaryEncoder encoder = EncoderFactory.get().directBinaryEncoder(out, null);
+            datumWriter.write(data, encoder);
+            encoder.flush();
+            return out.toByteArray();
+        } catch (IOException e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static <T> T deserialize(DatumReader<T> datumReader, byte[] data) {
+        try {
+            BinaryDecoder decoder = DecoderFactory.get().binaryDecoder(data, null);
+            return datumReader.read(null, decoder);
+        } catch (IOException e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static class GenericContainerHook implements SerializerHook<GenericContainer> {
+        private final Map<String, Schema> jsonToSchema = new HashMap<>();
+        private final Map<Schema, String> schemaToJson = new HashMap<>();
+
+        @Override
+        public Class<GenericContainer> getSerializationType() {
+            return GenericContainer.class;
+        }
+
+        @Override
+        public Serializer createSerializer() {
+            return new StreamSerializer<GenericContainer>() {
+                @Override
+                public int getTypeId() {
+                    return SerializerHookConstants.AVRO_GENERIC_CONTAINER;
+                }
 
-    public static final class Utf8Hook implements SerializerHook<Utf8> {
+                @Override
+                public void write(@Nonnull ObjectDataOutput out, @Nonnull GenericContainer datum) throws IOException {
+                    // SchemaNormalization.toParsingForm() drops details that are irrelevant to reader,
+                    // such as field default values. However:
+                    // 1. GenericData.Record.equals() compares record schemas without ignoring the dropped
+                    //    details, which is effectively schema1.toString().equals(schema2.toString()).
+                    // 2. There must be a 1-to-1 mapping between Schema's and their string representations
```

Comment:
> There must be a 1-to-1 mapping between Schema's and their string representations for bidirectional lookup.

Why? 

When you send data you convert Schema to JSON, with cache. Different schemas may generate the same JSON as long as this is correct for parsing (so if default values are irrelevant.

When you receive data you do not do a lookup in `schemaToJson` using JSON (map value). You instead parse JSON with separate cache (`jsonToSchema`).

The point about `GenericData.Record.equals()` still explains why do we use more verbose schema (toString instead of toParsingForm).

Reply:
The property mentioned in the second item is necessary to avoid collisions, but in this particular case, colliding schemas are equivalent for reading purposes, so I am removing it. As you pointed out, the actual reason why we cannot use an equivalent schema is explained in the first item.